assignments/introductory-assignments.md

```diff
@@ -488,14 +582,767 @@ A digital circuit computer with 2GB of memory can distinguish 2^{17 179 869 184}
+
+The mipster emulator implements a strict subset of 64-bit RISC-V instructions, which we call RISC-U. Mipster only implements 14 (32-bit-wide) out of several dozen RISC-V instructions. The starc compiler generates RISC-U code that runs on mipster and is compatible with the official RISC-V toolchain and runs, at least in principle, on real RISC-V processors.
+
+When talking about formal languages, it is essential to distinguish between the [syntax](https://en.wikipedia.org/wiki/Syntax) and the [semantics](https://en.wikipedia.org/wiki/Semantics) of that language.
+
+**Example** 
+
+In the above example, the `addi` string in `addi $a7,$zero,214` is an assembly mnemonic of the operation code or opcode, for short, that identifies the operation to be performed. In contrast, the remaining `$a7,$zero,214` are three operands of that operation.
+
+[Opcode](https://en.wikipedia.org/wiki/Opcode) The portion of a machine language instruction that specifies the operation to be performed. Besides the opcode itself, most instructions also specify the data they will process, in the form of operands.
+
+In our example, `addi` instructs the processor to add the value of the integer `214` to the value stored in register `$zero` (which is always 0) and store the result in register `$a7` (which will thus contain the value 214 after executing the instruction).
+
+A 64-bit RISC-V processor has 32 64-bit registers that can be used for this purpose. They are numbered from 0 to 31. The register `$zero` is the register 0 while the register `$a7`, less obviously, happens to be the register 17. The only missing piece of information is that `addi` is represented by the opcode 19. Now, how do we get from there to `0x0D600893`?
+
+We take:
+
+- the opcode 19 (`0010011`),
+- register 17 (`10001`),
+- register 0 (`00000`),
+- and value 214 (`000011010110`) 
+ 
+and put them together in binary as follows:
+
+`000011010110 00000 000 10001 0010011`
+
+If you merge that into a 32-bit word, you get `0x0D600893`.
+
+The RISC-V ISA specifies that:
+
+- the twelve MSBs encode the third (!) operand value 214, in 12-bit two's complement in fact, so it could even be a negative value.
+- The next five bits encode the second (!) operand register 0, followed by three bits set to 0.
+- The next five bits encode the first (!) operand register 17 followed by,
+- the remaining seven LSBs that encode the opcode 19.
+
+The third operand is treated by `addi` as an integer value rather than, say, a number identifying a register, is called *immediate* addressing hence the `i` in `addi`. Immediate addressing is one of various so-called *addressing modes*.
```

Comment:
```suggestion
The third operand is treated by `addi` as an integer value rather than, say, a number identifying a register, which is called immediate addressing hence the `i` in `addi`. Immediate addressing is one of the various so-called addressing modes.
```